tools/src/scie_pants/install_pants.py

```diff
@@ -19,54 +22,156 @@
+
+    def fetch_text(self, url: str, **headers: str) -> str:
+        return self._fetch(url, stdout=subprocess.PIPE, **headers).stdout.decode()
+
+    def fetch_to_fp(self, url: str, fp: BinaryIO, **headers: str) -> None:
+        self._fetch(url, stdout=fp.fileno(), **headers)
+
+
+@dataclass(frozen=True)
+class FindLinksRepo:
+    url: str
+
+    def iter_pip_options(self) -> Iterator[str]:
+        yield "--find-links"
+        yield self.url
+
+
+@dataclass(frozen=True)
+class ResolveInfo:
+    stable_version: Version
+    sha_version: Version
+    find_links: FindLinksRepo
+
+    def iter_pip_find_links_options(self) -> Iterator[str]:
+        yield from self.find_links.iter_pip_options()
+
+    def pants_find_links_option(self) -> str:
+        if self.stable_version in SpecifierSet("<2.14.0", prereleases=True):
+            return f"--python-repos-repos={self.find_links.url}"
+        return f"--python-repos-find-links={self.find_links.url}"
+
+
+def determine_find_links(
+    ptex: Ptex,
+    pants_version: str,
+    sha: str,
+    find_links_dir: Path,
+    include_pants_distributions_in_findlinks: bool,
+) -> ResolveInfo:
+    abbreviated_sha = sha[:8]
+    sha_version = Version(f"{pants_version}+git{abbreviated_sha}")
+
+    list_bucket_results = ElementTree.fromstring(
+        ptex.fetch_text(f"https://binaries.pantsbuild.org?prefix=wheels/3rdparty/{sha}")
+    )
+
+    find_links_file = find_links_dir / pants_version / abbreviated_sha / "index.html"
+    find_links_file.parent.mkdir(parents=True, exist_ok=True)
+    find_links_file.unlink(missing_ok=True)
+    with find_links_file.open("wb") as fp:
+        # N.B.: S3 bucket listings use a default namespace. Although the URI is apparently stable,
+        # we decouple from it with the wildcard.
+        for key in list_bucket_results.findall("./{*}Contents/{*}Key"):
+            bucket_path = str(key.text)
+            fp.write(
+                f'<a href="https://binaries.pantsbuild.org/{urllib.parse.quote(bucket_path)}">'
+                f"{os.path.basename(bucket_path)}"
+                f"</a>{os.linesep}".encode()
+            )
+        fp.flush()
+        if include_pants_distributions_in_findlinks:
+            pantsbuild_pants_find_links = (
+                "https://binaries.pantsbuild.org/wheels/pantsbuild.pants/"
+                f"{sha}/{urllib.parse.quote(str(sha_version))}/index.html"
+            )
+            ptex.fetch_to_fp(pantsbuild_pants_find_links, fp)
+
+    return ResolveInfo(
+        stable_version=Version(pants_version),
+        sha_version=sha_version,
+        find_links=FindLinksRepo(f"file://{find_links_file}"),
+    )
+
+
+def determine_tag_version(
+    ptex: Ptex, pants_version: str, find_links_dir: Path, github_api_bearer_token: str | None = None
+) -> ResolveInfo:
+    github_api_url = (
+        "https://api.github.com/repos/pantsbuild/pants/git/refs/tags/"
+        f"release_{urllib.parse.quote(pants_version)}"
+    )
+    headers = (
```

Comment:
Although this works, it's a bit titchy. It's OK if:
+ You have a ~/.netrc with an api.github.com entry (my personal case).
+ You export PANTS_BOOTSTRAP_GITHUB_API_BEARER_TOKEN (the case for this green CI)
+ You are a normal user (IP address) whose 60 unauthenticated request per hour quota is plenty (2 API calls per Pants version that needs to be installed).

It's not OK if your Pants using repo is hosted in, say, GitLab and you want to set up CI and you never go near GitHub. We shouldn't be asking you to create a GitHub account and generating a token to use just to use Pants - the two (should)have nothing to do with each other.

Mitigating all this is that this is just a bridge until Pants scie's start shipping (scie-pants does not install pants-scies right yet as you should notice).

It's easy enough for me to snapshot all tag -> sha in a json file say and embed that here. That though goes stale the minute we release 2.16.0.dev3, etc; so api requests are needed for those. I think the best path forward may be to backfill and start storing a https://binaries.pantsbuild.org/wheels/pantsbuild.pants/<tag> file in S3 as part of the release process for stable (tagged) releases. That would be a file that contains just the release tip commit sha. I could backfill those files for all existing tags and add that step to the currently active Pants release branches (main with cherry-picks back though 2.14.x).

Reply:
I agree on the release index (noting that using the GitHub API for supporting `PANTS_SHA` is probably fine long-term)

Reply:
Yeah, I think the index / tag files in S3 is the way to go. For the `PANTS_SHA` there is no `api.github.com` use and so no rate limit issue at all. We just scrape the raw VERSION file from `github.com` in that case.

Reply:
https://github.com/pantsbuild/scie-pants/blob/49399e7cc2c5fcd5f04d4a5441e64d063844bf21/tools/src/scie_pants/install_pants.py#L133-L140

Reply:
All right. The tag database is in now at ~43KB and ~30ms to load and lookup. That plus the work in https://github.com/pantsbuild/pants/pull/17801 and a one-time S3 back-fill will eliminate all GitHub API use.